File: tobas_gcs/src/tobas_gcs/apps/mission_planner/mission_execution_thread.py
import rclpy
from rclpy.duration import Duration
import actionlib
from typing import override
from typing import List
from PyQt5.QtCore import QObject, QThread, pyqtSignal

from tobas_tools_py.constants import Service, Action
from tobas_msgs.msg import (
    CommandLevel,
    TakeoffAction,
    TakeoffGoal,
    LandAction,
    LandGoal,
    MoveAction,
    MoveGoal,
)
from tobas_msgs.srv import GetGnssOrigin, GetGnssOriginRequest, GetGnssOriginResponse

from ...common import WAIT_FOR_SERVER
from .structs import *


class MissionExecutionThread(QThread):
    COMMAND_LEVEL = CommandLevel.NORMAL
    COMMAND_TIMEOUT = 10.0  # [s]  # TODO: ユーザが設定できるようにする
    WAIT_FOR_RESULT = 0.1  # [s]

    # 実行結果を返すためのシグナル．
    # NOTE: QThreadでGUIを使うとメインスレッドを壊す恐れがあるため，シグナルスロット以外は使用しない．
    finished = pyqtSignal(bool, str)

    def __init__(self, parent: QObject, drone_name: str, commands: List) -> None:
        super().__init__(parent)

        self._commands = commands
        self._stop_requested = False

        self._get_gnss_origin_sc = self._node.create_client(f"/{drone_name}/{Service.GET_GNSS_ORIGIN}", GetGnssOrigin)

        self._takeoff_ac = actionlib.SimpleActionClient(f"/{drone_name}/{Action.TAKEOFF}", TakeoffAction)
        self._land_ac = actionlib.SimpleActionClient(f"/{drone_name}/{Action.LAND}", LandAction)
        self._move_ac = actionlib.SimpleActionClient(f"/{drone_name}/{Action.MOVE}", MoveAction)

    @override
    def run(self) -> None:
        # 各サーバとの接続を確認
        if not self._check_server_connections():
            return

        # 前から順にコマンドを実行
        for command in self._commands:
            if not self._execute_command(command):
                return

        self.finished.emit(True, "")

    def stop(self) -> None:
        """
        実行中の関数に割り込んでミッションの終了フラグを立てる．

        NOTE
        ----------
        QThread.terminate()はクラッシュの恐れがあるため呼ばない．
        """
        self._stop_requested = True

    def _check_server_connections(self) -> bool:
        try:
            self._get_gnss_origin_sc.service_is_ready(WAIT_FOR_SERVER)
        except rclpy.ROSException:
            self.finished.emit(False, "Get GNSS origin server is not ready.")
            return False

        if not self._takeoff_ac.wait_for_server(Duration(WAIT_FOR_SERVER)):
            self.finished.emit(False, "Takeoff action server is not ready.")
            return False
        if not self._land_ac.wait_for_server(Duration(WAIT_FOR_SERVER)):
            self.finished.emit(False, "Land action server is not ready.")
            return False
        if not self._move_ac.wait_for_server(Duration(WAIT_FOR_SERVER)):
            self.finished.emit(False, "Move action server is not ready.")
            return False

        return True

    def _execute_command(self, command) -> bool:
        self.get_logger().info(f"Executing command: {command}")

        if isinstance(command, Waypoint):
            return self._execute_waypoint(command)
        elif isinstance(command, Takeoff):
            return self._execute_takeoff(command)
        elif isinstance(command, Land):
            return self._execute_land(command)
        elif isinstance(command, ReturnToHome):
            return self._execute_rth(command)
        else:
            raise RuntimeError(f"Unknown command type: {command.__class__.__name__}")

    def _execute_waypoint(self, command: Waypoint) -> bool:
        # ゴールを作成
        goal = MoveGoal()
        goal.level.data = self.COMMAND_LEVEL
        goal.target_latitude = command.latitude
        goal.target_longitude = command.longitude
        goal.target_altitude = command.altitude
        goal.acceptance_radius = command.acceptance_radius
        goal.duration = command.duration
        goal.timeout = self.COMMAND_TIMEOUT

        # アクションを実行
        self._move_ac.send_goal(goal)

        # 終了フラグを監視しながら待機
        while not self._move_ac.wait_for_result(Duration(self.WAIT_FOR_RESULT)):
            if self._stop_requested:
                self._move_ac.cancel_goal()
                return False

        if self._move_ac.get_state() != actionlib.GoalStatus.SUCCEEDED:
            self.finished.emit(False, self._move_ac.get_goal_status_text())
            return False

        return True

    def _execute_takeoff(self, command: Takeoff) -> bool:
        # ゴールを作成
        goal = TakeoffGoal()
        goal.level.data = self.COMMAND_LEVEL
        goal.target_altitude = command.altitude
        goal.altitude_tolerance = command.altitude_tolerance
        goal.duration = command.duration
        goal.timeout = self.COMMAND_TIMEOUT

        # アクションを実行
        self._takeoff_ac.send_goal(goal)

        # 終了フラグを監視しながら待機
        while not self._takeoff_ac.wait_for_result(Duration(self.WAIT_FOR_RESULT)):
            if self._stop_requested:
                self._takeoff_ac.cancel_goal()
                return False

        if self._takeoff_ac.get_state() != actionlib.GoalStatus.SUCCEEDED:
            self.finished.emit(False, self._takeoff_ac.get_goal_status_text())
            return False

        return True

    def _execute_land(self, command: Land) -> bool:
        # ゴールを作成
        goal = LandGoal()
        goal.level.data = self.COMMAND_LEVEL

        # アクションを実行
        self._land_ac.send_goal(goal)

        # 終了フラグを監視しながら待機
        while not self._land_ac.wait_for_result(Duration(self.WAIT_FOR_RESULT)):
            if self._stop_requested:
                self._land_ac.cancel_goal()
                return False

        if self._land_ac.get_state() != actionlib.GoalStatus.SUCCEEDED:
            self.finished.emit(False, self._land_ac.get_goal_status_text())
            return False

        return True

    def _execute_rth(self, command: ReturnToHome) -> bool:
        # ホームポジションの経緯度を取得
        res: GetGnssOriginResponse = self._get_gnss_origin_sc.call(GetGnssOriginRequest())
        if not res.success:
            self.finished.emit(False, f"Failed to get GNSS origin: {res.message}")

        # ゴールを作成
        goal = MoveGoal()
        goal.level.data = self.COMMAND_LEVEL
        goal.target_latitude = res.latitude
        goal.target_longitude = res.longitude
        goal.target_altitude = command.altitude
        goal.acceptance_radius = command.acceptance_radius
        goal.duration = command.duration
        goal.timeout = self.COMMAND_TIMEOUT

        # アクションを実行
        self._move_ac.send_goal(goal)

        # 終了フラグを監視しながら待機
        while not self._move_ac.wait_for_result(Duration(self.WAIT_FOR_RESULT)):
            if self._stop_requested:
                self._move_ac.cancel_goal()
                return False

        if self._move_ac.get_state() != actionlib.GoalStatus.SUCCEEDED:
            self.finished.emit(False, self._move_ac.get_goal_status_text())
            return False

        return True

File: tobas_gui/tobas_control_system/src/mission_planner/mission_execution_thread.cpp
#include "tobas_control_system/mission_planner/mission_execution_thread.hpp"

#include <boost/polymorphic_pointer_cast.hpp>

#include <tobas_path_tools/join.hpp>
#include <tobas_constants/constants.hpp>

using namespace std;
using namespace tobas_msgs::srv;
using namespace tobas_mission_msgs::action;

namespace gui
{
namespace gcs
{
MissionExecutionThread::MissionExecutionThread(rclcpp::Node::SharedPtr node) : node_(node)
{
}

void MissionExecutionThread::run()
{
  // Reset
  stop_requested_ = false;

  // Create service clients
  get_gnss_origin_sc_ = make_shared<ros2::SyncServiceClient<GetGnssOrigin>>(
    node_, path::join(ns_, tobas::kRemoteIfaceTopicNS, tobas::kGetGnssOriginSrv));

  // Create action clients
  takeoff_ac_ = make_shared<ros2::SyncActionClient<Takeoff>>(node_, path::join(ns_, tobas::kTakeoffAction));
  land_ac_ = make_shared<ros2::SyncActionClient<Land>>(node_, path::join(ns_, tobas::kLandAction));
  move_ac_ = make_shared<ros2::SyncActionClient<Move>>(node_, path::join(ns_, tobas::kMoveAction));

  // 前から順にコマンドを実行
  for (const auto& command : commands_) {
    if (!execute(command)) {
      return;
    }
  }

  Q_EMIT finished(true, "");
}

void MissionExecutionThread::setNamespace(const string& ns)
{
  ns_ = ns;
}

void MissionExecutionThread::setCommands(const QVector<BaseCommandData::SharedPtr>& commands)
{
  commands_ = commands;
}

void MissionExecutionThread::stop()
{
  stop_requested_ = true;
}

bool MissionExecutionThread::execute(BaseCommandData::SharedPtr command)
{
  const auto cmd_type = command->type();
  RCLCPP_INFO_STREAM(node_->get_logger(), "Execute command: " << commandToText(cmd_type));

  switch (cmd_type) {
    case command_t::WAYPOINT:
      return execute(boost::polymorphic_pointer_downcast<WaypointData>(command));
    case command_t::TAKEOFF:
      return execute(boost::polymorphic_pointer_downcast<TakeoffData>(command));
    case command_t::LAND:
      return execute(boost::polymorphic_pointer_downcast<LandData>(command));
    case command_t::RETURN_TO_HOME:
      return execute(boost::polymorphic_pointer_downcast<ReturnToHomeData>(command));
    default:
      Q_EMIT finished(false, format("Unknown command type: {}", (int)cmd_type).c_str());
      return false;
  }
}

bool MissionExecutionThread::execute(TakeoffData::SharedPtr command)
{
  // ゴールを作成
  Takeoff::Goal goal;
  goal.level.data = kCommandLevel;
  goal.target_altitude = command->altitude;
  goal.altitude_tolerance = command->altitude_tolerance;
  goal.duration = command->duration;
  goal.timeout = kCommandTimeout;

  // アクションを実行
  auto [goal_handle, get_result_future] = takeoff_ac_->sendGoal(goal);
  if (!get_result_future.valid()) {
    Q_EMIT finished(false, "Failed to call takeoff action.");
    return false;
  }

  // 終了フラグを監視しながら待機
  while (get_result_future.wait_for(kCheckCancelInterval) != std::future_status::ready) {
    if (stop_requested_) {
      takeoff_ac_->cancelGoal(goal_handle);
      return false;
    }
  }

  // アクションの成否を確認
  const auto result = get_result_future.get();
  if (result.code != rclcpp_action::ResultCode::SUCCEEDED) {
    Q_EMIT finished(false, result.result->message.c_str());
    return false;
  }

  return true;
}

bool MissionExecutionThread::execute(LandData::SharedPtr)
{
  // ゴールを作成
  Land::Goal goal;
  goal.level.data = kCommandLevel;

  // アクションを実行
  auto [goal_handle, get_result_future] = land_ac_->sendGoal(goal);
  if (!get_result_future.valid()) {
    Q_EMIT finished(false, "Failed to call land action.");
    return false;
  }

  // 終了フラグを監視しながら待機
  while (get_result_future.wait_for(kCheckCancelInterval) != std::future_status::ready) {
    if (stop_requested_) {
      land_ac_->cancelGoal(goal_handle);
      return false;
    }
  }

  // アクションの成否を確認
  const auto result = get_result_future.get();
  if (result.code != rclcpp_action::ResultCode::SUCCEEDED) {
    Q_EMIT finished(false, result.result->message.c_str());
    return false;
  }

  return true;
}

bool MissionExecutionThread::execute(WaypointData::SharedPtr command)
{
  // ゴールを作成
  Move::Goal goal;
  goal.level.data = kCommandLevel;
  goal.target_latitude = command->latitude;
  goal.target_longitude = command->longitude;
  goal.target_altitude = command->altitude;
  goal.acceptance_radius = command->acceptance_radius;
  goal.duration = command->duration;
  goal.timeout = kCommandTimeout;

  // アクションを実行
  auto [goal_handle, get_result_future] = move_ac_->sendGoal(goal);
  if (!get_result_future.valid()) {
    Q_EMIT finished(false, "Failed to call move action.");
    return false;
  }

  // 終了フラグを監視しながら待機
  while (get_result_future.wait_for(kCheckCancelInterval) != std::future_status::ready) {
    if (stop_requested_) {
      move_ac_->cancelGoal(goal_handle);
      return false;
    }
  }

  // アクションの成否を確認
  const auto result = get_result_future.get();
  if (result.code != rclcpp_action::ResultCode::SUCCEEDED) {
    Q_EMIT finished(false, result.result->message.c_str());
    return false;
  }

  return true;
}

bool MissionExecutionThread::execute(ReturnToHomeData::SharedPtr command)
{
  // ホームポジションの経緯度を取得
  const auto get_gnss_origin_req = make_shared<tobas_msgs::srv::GetGnssOrigin::Request>();
  if (!get_gnss_origin_sc_->call(get_gnss_origin_req)) {
    Q_EMIT finished(false, format("Failed to call \"{}\" service.", tobas::kGetGnssOriginSrv).c_str());
    return false;
  }

  const auto get_gnss_origin_res = get_gnss_origin_sc_->getResponse();
  if (!get_gnss_origin_res->success) {
    Q_EMIT finished(false, format("Failed to get GNSS origin: {}", get_gnss_origin_res->message).c_str());
    return false;
  }

  // ゴールを作成
  Move::Goal goal;
  goal.level.data = kCommandLevel;
  goal.target_latitude = get_gnss_origin_res->latitude;
  goal.target_longitude = get_gnss_origin_res->longitude;
  goal.target_altitude = command->altitude;
  goal.acceptance_radius = command->acceptance_radius;
  goal.duration = command->duration;
  goal.timeout = kCommandTimeout;

  // アクションを実行
  auto [goal_handle, get_result_future] = move_ac_->sendGoal(goal);
  if (!get_result_future.valid()) {
    Q_EMIT finished(false, "Failed to call move action.");
    return false;
  }

  // 終了フラグを監視しながら待機
  while (get_result_future.wait_for(kCheckCancelInterval) != std::future_status::ready) {
    if (stop_requested_) {
      move_ac_->cancelGoal(goal_handle);
      return false;
    }
  }

  // アクションの成否を確認
  const auto result = get_result_future.get();
  if (result.code != rclcpp_action::ResultCode::SUCCEEDED) {
    Q_EMIT finished(false, result.result->message.c_str());
    return false;
  }

  return true;
}
}  // namespace gcs
}  // namespace gui
